File: Combination Sum II.cpp
#include <iostream>
#include <vector>

class Solution {
  public:
    std::vector<std::vector<int>> combinationSum2(std::vector<int> &candidates,
                                                  int target) {
        std::sort(candidates.begin(), candidates.end());
        std::vector<std::vector<int>> res;
        std::vector<int> cur;
        dfs(candidates, target, res, cur, 0, 0);

        return res;
    }

    void dfs(std::vector<int> &candidates, int target,
             std::vector<std::vector<int>> &res, std::vector<int> &cur, int sum,
             int index) {
        if (sum > target) {
            return;
        }
        if (sum == target) {
            res.push_back(cur);
            return;
        }

        for (int i = index; i < candidates.size(); ++i) {
            std::cout << index << ' ' << i << std::endl;
            if (i > index && candidates[i] == candidates[i - 1]) {
                continue;
            }
            cur.push_back(candidates[i]);
            dfs(candidates, target, res, cur, sum + candidates[i], i + 1);
            cur.pop_back();
        }
    }
};

File: Combination Sum II.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Solution {
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Arrays.sort(candidates);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> cur = new ArrayList<>();
        dfs(candidates, target, res, cur, 0, 0);

        return res;
    }

    private void dfs(int[] candidates, int target, List<List<Integer>> res,
                     List<Integer> cur, int sum, int index) {
        if (sum > target) {
            return;
        }
        if (sum == target) {
            res.add(new ArrayList<>(cur));
            return;
        }

        for (int i = index; i < candidates.length; ++i) {
            if (i > index && candidates[i] == candidates[i - 1]) {
                continue;
            }
            cur.add(candidates[i]);
            dfs(candidates, target, res, cur, sum + candidates[i], i + 1);
            cur.remove(cur.size() - 1);
        }
    }
}
